### PR TITLE
Time Validity PR

### PR DESCRIPTION
**Pull request by rgottsc**
Time validity of a dataset is now displayed

### CODE REVIEW

**Comment by rgottsc** [comment]
#83

**Comment by ChristianBeilschmidt** [comment]
[Time validity indicator in data repository](https://trello.com/c/DcDEuweR/1-time-validity-indicator-in-data-repository)

File: projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts [hunk 1]
```diff
@@ -102,6 +109,8 @@ export interface MappingSourceDict {
     };
     channels?: [MappingSourceRasterLayerDict];
     layer?: [MappingSourceVectorLayerDict];
+    time_start?: Time;
+    time_end?: Time;
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts:113]: The source delivers strings and not `Time` types I guess. This is why you don't construct these types but just assign them. But in the end, `toString` of `Time` is not called. You see this in the different output format.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts [hunk 2]
```diff
@@ -131,6 +140,8 @@ export interface MappingSourceRasterLayerDict {
         license: string,
         citation: string,
     };
+    time_start?: Time;
+    time_end?: Time;
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts:144]: Same here. `…Dict`s always reflect the serialized forms.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts [hunk 3]
```diff
@@ -154,6 +165,8 @@ export interface MappingSourceVectorLayerDict {
...
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts:169]: Same here.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts [hunk 4]
```diff
@@ -34,7 +34,7 @@ export class SourceDatasetComponent implements OnInit {
     _showPreview = false;
     _showDescription = false;
     _channelSource: ChannelDataSource;
-    _displayedColumns = ['name', 'measurement'];
+    _displayedColumns = ['name', 'measurement', 'start', 'end'];
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts:37]: You should display `start` and `end` only if there is information available.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts [hunk 5]
```diff
@@ -33,7 +33,7 @@ export class VectorSourceDatasetComponent implements OnInit {
     _showPreview = false;
     _showDescription = false;
     _tableSource: LayerTableDataSource;
-    _displayedColumns  = ['title'];
+    _displayedColumns  = ['title', 'start', 'end'];
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:36]: Same here.

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: projects/wave-core/src/lib/users/user.service.ts [hunk 6]
```diff
@@ -477,6 +480,8 @@ export class UserService {
             geometryType: layer.geometry_type,
             textual: layer.textual || [],
             numeric: layer.numeric || [],
+            time_start: !!layer.time_start ? new TimePoint(layer.time_start) : new TimePoint('') ,
+            time_end: !!layer.time_end ? new TimePoint(layer.time_end) : new TimePoint('') ,
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/users/user.service.ts:484]: Both fields are marked as optional on the type `SourceVectorLayerDescription`, so it could just be undefined if the field is not present.

File: projects/wave-core/src/lib/users/user.service.ts [hunk 7]
```diff
@@ -516,6 +521,8 @@ export class UserService {
             } as MappingTransform,
             coords: coords as { crs: string, origin: number[], scale: number[], size: number[] },
             provenance: channelProvenance,
+            time_start: !!channel.time_start ? new TimePoint(channel.time_start) : new TimePoint('') ,
+            time_end: !!channel.time_end ? new TimePoint(channel.time_end) : new TimePoint('') ,
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/users/user.service.ts:525]: Same here.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts [hunk 8]
```diff
@@ -25,7 +25,7 @@ import {NotificationService} from '../../../../notification.service';
     changeDetection: ChangeDetectionStrategy.OnPush,
 })
 
-export class VectorSourceDatasetComponent implements OnInit {
+export class VectorSourceDatasetComponent implements OnInit , OnChanges  {
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:28]: Please format the code.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts [hunk 9]
```diff
@@ -26,15 +26,15 @@ import {Colormap} from '../../../../colors/colormaps/colormap.model';
     changeDetection: ChangeDetectionStrategy.OnPush,
 })
 
-export class SourceDatasetComponent implements OnInit {
+export class SourceDatasetComponent implements OnInit , OnChanges {
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts:29]: Same here.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts [hunk 10]
```diff
@@ -26,15 +26,15 @@ import {Colormap} from '../../../../colors/colormaps/colormap.model';
...
 
     @Input() dataset: MappingSource;
     @Input() searchTerm: string;
     _useRawData = false;
     _showPreview = false;
     _showDescription = false;
     _channelSource: ChannelDataSource;
-    _displayedColumns = ['name', 'measurement'];
+    _displayedColumns = ['name', 'measurement'] ;
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts:37]: Please revert this change.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts [hunk 11]
```diff
@@ -96,6 +96,68 @@ export class VectorSourceDatasetComponent implements OnInit {
...
+                switch (key) {
+                    // check if there is any time-validity start/end data. If there is start/end data show the column of this data.
+                    // Also find a global Start/End (Min/Max) for the whole dataset.
+                    case 'dataset': {
+                        let globalDatasetMin = this.dataset.time_start;
+                        let globalDatasetMax = this.dataset.time_end;
+                        this.dataset.vectorLayer.forEach((element) => {
+                            for (const sourceVar in element) {
+                                if (element.hasOwnProperty(sourceVar)) {
+                                    if (sourceVar === 'time_start') {
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:112]: You know the type of `element`, so you can access the fields directly without needing to iterate over them, check field names, etc.

Comment (ChristianBeilschmidt) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:112]: Its the same for the checks in source-dataset.component.ts

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts [hunk 12]
```diff
@@ -96,6 +96,68 @@ export class VectorSourceDatasetComponent implements OnInit {
...
+                                        if (!element.time_end.isValid()) {
+                                            element.time_end = globalDatasetMax;
+                                        }
+                                    }
+                                }
+                            }
+                        });
+                        // Update the global Min/Max in the whole dataset
+                        this.dataset.time_start = globalDatasetMin;
+                        this.dataset.time_end = globalDatasetMax;
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:154]: If you change the dataset, then move this processing to the `UserService`.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts [hunk 13]
```diff
@@ -96,6 +96,68 @@ export class VectorSourceDatasetComponent implements OnInit {
...
+                                }
+                            }
+                        });
+                        // Fill in the global Min/Max for rows without Start/End
+                        this.dataset.vectorLayer.forEach((element) => {
+                            for (const sourceVar in element) {
+                                if (element.hasOwnProperty(sourceVar)) {
+                                    if (sourceVar === 'time_start') {
+                                        if (!element.time_start.isValid()) {
+                                            element.time_start = globalDatasetMin;
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:141]: This is unfortunately incorrect. The rule is: If the layer has no own time interval, use the one from the dataset.
If you changed the dataset's validity before to display the global min/max, then this potentially produces wrong intervals.

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]
I have some comments regarding the computations. Furthermore, please address my issues one by one (also the ones before). Otherwise, I can't resolve them to see if I can merge the PR.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts (hunk 1 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts]: fixed in 9919298afbd39ba8285c6946f1eec93e1c65b5ba

File: projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts (hunk 2 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts]: fixed in 9919298afbd39ba8285c6946f1eec93e1c65b5ba

File: projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts (hunk 3 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/mapping-source.model.ts]: fixed in 9919298afbd39ba8285c6946f1eec93e1c65b5ba

File: projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts (hunk 4 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts]: fixed in bc0236c100809397960ca46b0f1e76b973d72473

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts (hunk 5 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts]: fixed in bc0236c100809397960ca46b0f1e76b973d72473

File: projects/wave-core/src/lib/users/user.service.ts (hunk 6 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/users/user.service.ts]: fixed in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

File: projects/wave-core/src/lib/users/user.service.ts (hunk 7 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/users/user.service.ts]: fixed in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts (hunk 8 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts]: formated in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f****

File: projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts (hunk 9 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts]: formated in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

File: projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts (hunk 10 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.ts]: whitespaced deleted in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts (hunk 11 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts]: deleted the useless for-loop in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts (hunk 13 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts]: i hope its fixed in 6da1d4c74b7a75a702074ebf5ef83c4bda092d4f

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts (hunk 12 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts]: reverted the dataset manipulation/processing part in bcf99b0df6ed19d67ce802b25df192cf7edd124e

As discussed i would like to do the dataset processing part in a new PR

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts (hunk 13 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts]: reverted the dataset manipulation/processing part in bcf99b0df6ed19d67ce802b25df192cf7edd124e

As discussed i would like to do the dataset processing part in a new PR

File: projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.html [hunk 14]
```diff
@@ -16,12 +16,16 @@
     </mat-chip-list>
 
     <mat-list>
+        <mat-list-item *ngIf="dataset?.time_start || dataset?.time_end">
+            <p><b>Time validity:</b> {{dataset?.time_start.toString()}} - {{dataset?.time_end.toString()}}</p>
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.html:20]:
```suggestion
            <p><b>Time validity:</b> {{dataset?.time_start}} - {{dataset?.time_end}}</p>
```
I think `toString()` is called automatically in templates, so it is redundant and you do it differently below.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.html [hunk 15]
```diff
@@ -12,6 +12,9 @@
     </mat-card-content>
 
     <mat-list>
+        <mat-list-item *ngIf="dataset?.time_start || dataset?.time_end">
+            <p><b>Time validity:</b> {{dataset?.time_start.toString()}} - {{dataset?.time_end.toString()}}</p>
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.html:16]:
```suggestion
            <p><b>Time validity:</b> {{dataset?.time_start}} - {{dataset?.time_end}}</p>
```

same here

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts [hunk 16]
```diff
@@ -96,6 +96,29 @@ export class VectorSourceDatasetComponent implements OnInit {
...
         this.projectService.addLayer(l);
     }
+    ngOnChanges(changes: SimpleChanges) {
+        for (const key in changes) {
+            if (changes.hasOwnProperty(key)) {
+                switch (key) {
+                    // check if there is any time-validity start/end data. If there is start/end data show the column of this data.
+                    case 'dataset': {
+                        this.dataset.rasterLayer.forEach((element) => {
+                            if (element.time_start.isValid()) {
```
Comment (ChristianBeilschmidt) [inline projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts:106]: What happens if `time_start` is undefined?

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]
There are two minor things and one potential bug. The rest looks fine.

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts (hunk 16 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.ts]: fixed in 1a0b23afc2302b363ceb11f273d325f130d8539d

File: projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.html (hunk 15 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/vector/vector-source-dataset.component.html]: fixed in 900ce5db69b7712d277d6400055ede30c92368c1

File: projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.html (hunk 14 above)
Comment (rgottsc) [reply projects/wave-core/src/lib/operators/dialogs/data-repository/raster/source-dataset.component.html]: fixed in 900ce5db69b7712d277d6400055ede30c92368c1

**Comment by ChristianBeilschmidt** [comment]
grafik

Can you look at these datasets and figure out, why the wrong time validity (now) is shown?

**Comment by rgottsc** [comment]
Unfortunately, I did not implement the undefined case correctly at one point in `user-service.ts`. It is now fixed. Furthermore I changed the ng-if in `source-dataset.component.htm`l and `vector-source-dataset.comonent.html`  a little bit, so that the time validity is only shown if start and end are not undefined

**Review by ChristianBeilschmidt — APPROVED** [review]
